### DISCUSSION TITLE
Resolving Compilation Issue with compilation in Oqtane 5.1.0 when copying source files over existing version.

### DISCUSSION THREAD

**Discussion by leigh-pointer** [oqtane/oqtane.framework#4076] [General]
I recently encountered an issue while working with Oqtane 5.1.0, where the compilation process failed due to an ambiguous reference error related to 'Runtime'. Upon further investigation, I discovered that the problem arises when copying source files over an existing version of Oqtane.

However, I'm glad to share a solution that I found to resolve this issue. It requires a manual deletion of certain files within the Oqtane project directory. By removing the following files, the compilation error can be successfully mitigated:

- Oqtane.Client\App.razor
- Oqtane.Client\UI\Runtime.cs
- Oqtane.Client\UI\PaneLayout.razor
- Oqtane.Client\Modules\Admin\Themes\Themes\View.razor
- Oqtane.Server\Pages_Host.cshtml
- Oqtane.Server\Pages_Host.cshtml.cs
- Oqtane.Server\Infrastructure\HostedServices\CacheInvalidationHostedService.cs
- Oqtane.Server\Infrastructure\SiteState.cs

While this solution might seem straightforward, it's crucial to ensure that you have a backup of your project files before proceeding with any manual deletions. Additionally, always verify that the files you're removing are indeed the ones causing the compilation error.

By implementing this solution, you should be able to compile your Oqtane 5.1.0 project without encountering the ambiguous reference error to 'Runtime'. However, if you face any challenges or have alternative solutions to share, I encourage you to contribute to this discussion.

**Comment by sbwalker** [comment]
One other item you might run into in your custom modules when you change your Oqtane package reference to 5.1.0 is a compilation error:

```
The type 'IHttpClientFactory' is defined in an assembly that is not referenced. You must add a reference to assembly 'Microsoft.Extensions.Http, Version=8.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'
```

This is because the ServiceBase constructor was enhanced to support IHttpClientFactory as well as the legacy HttpClient approach. And even if you are not using the new IHttpClientFactory service class in your Services, the compiler still needs a package reference for it. This can easily be resolved by adding the following to your Client project:

```
    <PackageReference Include="Microsoft.Extensions.Http" Version="8.0.0" />
```

Note that modules which are compiled against older Oqtane versions are still compatible from a run-time perspective.
